### DISCUSSION TITLE
Is there a way to retry image loading on failure using Glide?

### DISCUSSION THREAD

**Discussion by stenol53** [skydoves/landscapist#453] [Q&A]
I have a case where the server sends me the url of an image, but the image itself is not ready yet, and a request for this url may return a 404 error. So I need to automatically retry image loading several times with some delay before showing error

**Comment by faraz152** [comment]
Glide doesn't have a built-in "retry N times with delay" feature, but you can implement it cleanly in Compose by tracking a retry key in state. When the image fails to load, incrementing the key forces landscapist to restart the request.

### Simple retry-on-failure pattern

```kotlin
@Composable
fun RetryableGlideImage(url: String, maxRetries: Int = 3) {
    var retryKey by remember { mutableIntStateOf(0) }
    var retryCount by remember { mutableIntStateOf(0) }

    // Re-trigger with a delay when retryKey changes (except the first load)
    LaunchedEffect(retryKey) {
        if (retryKey > 0) {
            delay(1_000L * retryKey) // exponential-ish back-off: 1s, 2s, 3s...
        }
    }

    GlideImage(
        imageModel = { "$url?retry=$retryKey" }, // cache-busts the request on retry
        modifier = Modifier.fillMaxWidth(),
        failure = {
            if (retryCount < maxRetries) {
                LaunchedEffect(Unit) {
                    retryCount++
                    delay(1_000L * retryCount)
                    retryKey++
                }
            } else {
                // Show a permanent error placeholder
                Text("Failed to load after $maxRetries attempts")
            }
        }
    )
}
```

### Bypass the disk cache on retries

If the server returns a 404 while the image isn't ready yet, Glide might cache that failure. Force a fresh network request by applying `DiskCacheStrategy.NONE` on retries:

```kotlin
GlideImage(
    imageModel = { url },
    requestBuilder = { requestBuilder ->
        if (retryCount > 0) {
            requestBuilder.diskCacheStrategy(DiskCacheStrategy.NONE)
                         .skipMemoryCache(true)
        } else {
            requestBuilder
        }
    },
    failure = { /* trigger retry as above */ }
)
```

### Key points

- The `failure` slot in landscapist fires whenever Glide's request fails (4xx, 5xx, network error).
- Changing the `imageModel` lambda's return value is enough to restart the load — no need to remove/re-add the composable.
- Use `diskCacheStrategy(DiskCacheStrategy.NONE)` + `skipMemoryCache(true)` when the URL is temporarily returning 404 so Glide doesn't serve the cached failure.
